File: PayMeBack.Backend.Services/ContactService.cs
using System.Collections.Generic;
using PayMeBack.Backend.Models;
using PayMeBack.Backend.Contracts.Services;
using PayMeBack.Backend.Contracts;
using System;
using System.Linq;

namespace PayMeBack.Backend.Services
{
    public class ContactService : IContactService
    {
        private IGenericRepository<Contact> _contactRepository;
        private IGenericRepository<SplitContact> _splitContactRepository;

        public ContactService(IGenericRepository<Contact> contactRepository, IGenericRepository<SplitContact> splitContactRepository)
        {
            _contactRepository = contactRepository;
            _splitContactRepository = splitContactRepository;
        }

        public IEnumerable<SplitContact> ListSplitContactsBySplitId(int splitId)
        {
            var splitContacts = _splitContactRepository.GetWithIncludedProperties(sc => sc.Contact, sc => sc.SplitId == splitId);

            return splitContacts;
        }

        public Contact CreateIfNeededAndAddToSplit(int splitId, string email, string name)
        {
            var contact = _contactRepository.Get(c => c.Email == email).FirstOrDefault();

            if (contact == null)
            {
                var nameToUse = string.IsNullOrEmpty(name) ? email : name;
                contact = new Contact { Name = nameToUse, Email = email };
                _contactRepository.Insert(contact);
            }

            _splitContactRepository.Insert(new SplitContact { SplitId = splitId, ContactId = contact.Id });
            
            return contact;
        }

        public Contact GetContactById(int contactId)
        {
            return _contactRepository.Get(c => c.Id == contactId).FirstOrDefault();
        }

        public SplitContact GetSplitContactById(int splitContactId)
        {
            return _splitContactRepository.Get(sc => sc.Id == splitContactId).FirstOrDefault();
        }

        public void UpdateSplitContact(int splitContactId, decimal owes, decimal paid, string iban, string address, string comments)
        {
            var splitContact = _splitContactRepository.GetWithIncludedProperties(sc => sc.Contact, sc => sc.Id == splitContactId).FirstOrDefault();

            splitContact.Owes = owes;
            splitContact.Paid = paid;
            splitContact.Contact.Iban = iban;
            splitContact.Contact.Address = address;
            splitContact.Comments = comments;

            _splitContactRepository.Update(splitContact);
        }
    }
}

File: PayMeBack.Frontend/www/scripts/services/contactService.js
angular.module('PayMeBack').service('contactService', ['backendHostUrl', '$http', '$cordovaContacts', '$ionicPlatform', contactService]);

function contactService(backendHostUrl, $http, $cordovaContacts, $ionicPlatform) {


    this.getBySplitId = function (splitId) {
        return $http.get(backendHostUrl + '/splits/' + splitId + '/contacts').then(
            function successCallback(response) {
                return response.data;
            }, _errorCallback);
    };

    this.createIfNeededAndAddToSplit = function (splitId, contactEmailToAdd, contactNameToAdd) {
        var contactToCreate = { email: contactEmailToAdd, splitId: splitId, name: contactNameToAdd };
        return $http.post(backendHostUrl + '/splits/' + splitId + '/contacts', contactToCreate).then(
            function successCallback(response) {
                return response.data;
            }, _errorCallback);
    };

    this.getSplitContactById = function (splitId, splitContactId) {
        return $http.get(backendHostUrl + '/splits/' + splitId + '/contacts/' + splitContactId).then(
            function successCallback(response) {
                return response.data;
            }, _errorCallback);
    };

    this.updateSplitContact = function (splitContact) {
        var data = { owes: splitContact.owes, paid: splitContact.paid, iban: splitContact.iban, address: splitContact.address, comments: splitContact.comments };
        return $http.put(backendHostUrl + '/splits/' + splitContact.splitId + '/contacts/' + splitContact.id, data).then(
            function successCallback(response) {
                return response.data;
            }, _errorCallback);
    };

    this.searchPhoneContacts = function (searchTerm) {
        // Used for debugging on Chrome
        //return { then: function (callback) { callback([{ displayName: 'Olivier Roger', emails: [{ value: 'olivier.roger@gmail.com' }] }, { displayName: 'Olivier Desvachez', emails: [{ value: 'olivier.desvachez@gmail.com' }] }]) } };

        //$ionicPlatform.ready(function () { // disabled because it doesn't return a promise
        var options = {
            filter: searchTerm, // 'Bob'
            multiple: true, // Yes, return any contact that matches criteria.
            fields: ['displayName', 'name', 'emails'], // These are the fields to search for 'bob'.
            desiredFields: ['displayName', 'name', 'emails'] //return fields, others will be null.
        };
        return $cordovaContacts.find(options);
        //});
    };

    function _errorCallback(response) {
        return $q.reject();
    }

}